libdnf/rpm/repo.cpp

```diff
@@ -0,0 +1,2058 @@
+    }
+    if (!conf->sslclientkey().get_value().empty()) {
+        handle_set_opt(h.get(), LRO_SSLCLIENTKEY, conf->sslclientkey().get_value().c_str());
+    }
+
+    handle_set_opt(h.get(), LRO_PROGRESSCB, static_cast<LrProgressCb>(progress_cb));
+    handle_set_opt(h.get(), LRO_PROGRESSDATA, callbacks.get());
+    handle_set_opt(h.get(), LRO_FASTESTMIRRORCB, static_cast<LrFastestMirrorCb>(fastest_mirror_cb));
+    handle_set_opt(h.get(), LRO_FASTESTMIRRORDATA, callbacks.get());
+
+#ifdef LRO_SUPPORTS_CACHEDIR
+    // If zchunk is enabled, set librepo cache dir
+    if (conf->getMasterConfig().zchunk().get_value()) {
+        handleSetOpt(h.get(), LRO_CACHEDIR, conf->basecachedir().get_value().c_str());
+    }
+#endif
+
+    auto minrate = conf->minrate().get_value();
+    handle_set_opt(h.get(), LRO_LOWSPEEDLIMIT, static_cast<long>(minrate));
+
+    auto maxspeed = conf->throttle().get_value();
+    if (maxspeed > 0 && maxspeed <= 1) {
+        maxspeed *= static_cast<float>(conf->bandwidth().get_value());
+    }
+    if (maxspeed != 0 && maxspeed < static_cast<float>(minrate)) {
+        throw RuntimeError(
+            _("Maximum download speed is lower than minimum. "
+              "Please change configuration of minrate or throttle"));
+    }
+    handle_set_opt(h.get(), LRO_MAXSPEED, static_cast<int64_t>(maxspeed));
+
+    long timeout = conf->timeout().get_value();
+    if (timeout > 0) {
+        handle_set_opt(h.get(), LRO_CONNECTTIMEOUT, timeout);
+        handle_set_opt(h.get(), LRO_LOWSPEEDTIME, timeout);
+    } else {
+        handle_set_opt(h.get(), LRO_CONNECTTIMEOUT, LRO_CONNECTTIMEOUT_DEFAULT);
+        handle_set_opt(h.get(), LRO_LOWSPEEDTIME, LRO_LOWSPEEDTIME_DEFAULT);
+    }
+
+    if (!conf->proxy().empty() && !conf->proxy().get_value().empty()) {
+        handle_set_opt(h.get(), LRO_PROXY, conf->proxy().get_value().c_str());
+    }
+
+    //set proxy authorization method
+    auto proxy_auth_method_str = conf->proxy_auth_method().get_value();
+    auto proxy_auth_method = LR_AUTH_ANY;
+    for (auto & auth : PROXYAUTHMETHODS) {
+        if (proxy_auth_method_str == auth.name) {
+            proxy_auth_method = auth.code;
+            break;
+        }
+    }
+    handle_set_opt(h.get(), LRO_PROXYAUTHMETHODS, static_cast<long>(proxy_auth_method));
+
+    if (!conf->proxy_username().empty()) {
+        userpwd = conf->proxy_username().get_value();
+        if (!userpwd.empty()) {
+            userpwd = format_user_pass_string(userpwd, conf->proxy_password().get_value(), true);
+            handle_set_opt(h.get(), LRO_PROXYUSERPWD, userpwd.c_str());
+        }
+    }
+
+    auto sslverify = conf->sslverify().get_value() ? 1L : 0L;
+    handle_set_opt(h.get(), LRO_SSLVERIFYHOST, sslverify);
+    handle_set_opt(h.get(), LRO_SSLVERIFYPEER, sslverify);
+
+    return h;
+}
+
+static void gpg_import_key(gpgme_ctx_t context, int key_fd, Logger & logger) {
```

Comment:
Please could you add TODO - replace gpgme by something else?

Reply:
Done.